api/migrations/20200617203111-roleUpdates.js

```diff
@@ -0,0 +1,117 @@
+'use strict';
+
+const RECORD_TYPE = require('../src/utils/constants/record-type-enum');
+
+let dbm;
+let type;
+let seed;
+
+const roles = {
+  ADMIN:       'sysadmin',
+  ADMIN_NRCED: 'admin:nrced',
+  ADMIN_LNG:   'admin:lng',
+  ADMIN_BCMI:  'admin:bcmi'
+};
+
+const ALL_ROLES = [roles.ADMIN,
+                   roles.ADMIN_NRCED,
+                   roles.ADMIN_LNG,
+                   roles.ADMIN_BCMI];
+
+/**
+  * We receive the dbmigrate dependency from dbmigrate initially.
+  * This enables us to not have to rely on NODE_PATH.
+  */
+exports.setup = function(options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+exports.up = async function (db) {
+  let mClient;
+
+  try {
+    mClient = await db.connection.connect(db.connectionString, { native_parser: true });
+    const nrpti = mClient.collection('nrpti');
+
+    // Master documents get all roles
+    // flavour documents get their specific flavour, right now just lng and nrced
+
+    console.log('-------------------------------');
+    console.log('     Starting Role Updates');
+    console.log('-------------------------------');
+    console.log('\nAdding the following roles to MASTER WRITE: ' + ALL_ROLES);
+    console.log('Adding the following roles to LNG WRITE:    ' + [roles.ADMIN, roles.ADMIN_LNG]);
+    console.log('Adding the following roles to NRCED WRITE:  ' + [roles.ADMIN, roles.ADMIN_NRCED]);
+    console.log('\nAdding the following roles to MASTER READ: ' + [roles.ADMIN_LNG, roles.ADMIN_NRCED, roles.ADMIN_BCMI]);
+    console.log('Adding the following roles to LNG READ:    ' + [roles.ADMIN_LNG, roles.ADMIN_NRCED, roles.ADMIN_BCMI]);
+    console.log('Adding the following roles to NRCED READ:  ' + [roles.ADMIN_LNG, roles.ADMIN_NRCED, roles.ADMIN_BCMI] + '\n');
+
+    let updateResults = await documentUpdates(nrpti);
+
+    let found = 0;
+    let modified = 0;
+    let failed = 0;
+
+    updateResults.forEach(update => {
+      found += update.result.n;
+      modified += update.result.nModified;
+      failed += update.result.ok !== 1 ? 1 : 0;
+    });
+
+    console.log(`\nQueried ${found} documents, updated ${modified}`);
+    console.log(`${updateResults.length} queries executed, ${failed} failures.`);
+  } catch(err) {
+    console.log('\n// ###############################');
+    console.error('// Error on Role Updates: ' + err);
+    console.log('// ###############################\n');
+  }
+
+  console.log('\n-------------------------------');
+  console.log('     Role Updates Complete');
+  console.log('-------------------------------');
+
+  mClient.close();
+};
+
+exports.down = function(db) {
+  return null;
+};
+
+exports._meta = {
+  "version": 1
+};
+
+async function documentUpdates(nrpti) {
+  let promises = [];
+
+  const masterUpdate = { $set: { write: ALL_ROLES }, $push: { read: { $each: [roles.ADMIN_LNG, roles.ADMIN_NRCED, roles.ADMIN_BCMI] }}};
```

Comment:
Should the read array be explicitly set just in case there is some bad data in there or if sysadmin is missing? Right now it is just appending to whatever already exists.

Reply:
Good catch. Changing to `addToSet`

Reply:
Follow up note: The reason I'm not explicitly setting the read arrays is so I don't change any existing `public` roles. I could do that with a static set, but I'd have to check the array first, push/addToSet ensures we inject only the new roles without altering the existing roles, valid or otherwise. `$addToSet` is better then `$push` in this case because we don't want dupes on a re-run of the migration.